api-samples/downloads/download_manager/background.js

```diff
@@ -0,0 +1,267 @@
+    ctx.fillStyle = ctx.strokeStyle = (
+      ((seg % 2) == 0) ? colors.progressColor : colors.background);
+    drawProgressArc(ctx, (seg-4)*segArc, (seg-3)*segArc);
+  }
+}
+
+function drawProgressSpinner(ctx, stage) {
+  ctx.fillStyle = ctx.strokeStyle = colors.progressColor;
+  var clocktop = -Math.TAU/4;
+  drawProgressArc(ctx, clocktop, clocktop + (stage * Math.TAU));
+}
+
+function drawArrow(ctx) {
+  ctx.beginPath();
+  ctx.lineWidth = Math.round(ctx.canvas.width*0.1);
+  ctx.lineJoin = 'round';
+  ctx.strokeStyle = ctx.fillStyle = colors.arrow;
+  var center = ctx.canvas.width/2;
+  var minw2 = center*0.2;
+  var maxw2 = center*0.60;
+  var height2 = maxw2;
+  ctx.moveTo(center-minw2, center-height2);
+  ctx.lineTo(center+minw2, center-height2);
+  ctx.lineTo(center+minw2, center);
+  ctx.lineTo(center+maxw2, center);
+  ctx.lineTo(center, center+height2);
+  ctx.lineTo(center-maxw2, center);
+  ctx.lineTo(center-minw2, center);
+  ctx.lineTo(center-minw2, center-height2);
+  ctx.lineTo(center+minw2, center-height2);
+  ctx.stroke();
+  ctx.fill();
+}
+
+function drawDangerBadge(ctx) {
+  var s = ctx.canvas.width/100;
+  ctx.fillStyle = colors.danger;
+  ctx.strokeStyle = colors.background;
+  ctx.lineWidth = Math.round(s*5);
+  var edge = ctx.canvas.width-ctx.lineWidth;
+  ctx.beginPath();
+  ctx.moveTo(s*75, s*55);
+  ctx.lineTo(edge, edge);
+  ctx.lineTo(s*55, edge);
+  ctx.lineTo(s*75, s*55);
+  ctx.lineTo(edge, edge);
+  ctx.fill();
+  ctx.stroke();
+}
+
+function drawPausedBadge(ctx) {
+  var s = ctx.canvas.width/100;
+  ctx.beginPath();
+  ctx.strokeStyle = colors.background;
+  ctx.lineWidth = Math.round(s*5);
+  ctx.rect(s*55, s*55, s*15, s*35);
+  ctx.fillStyle = colors.paused;
+  ctx.fill();
+  ctx.stroke();
+  ctx.rect(s*75, s*55, s*15, s*35);
+  ctx.fill();
+  ctx.stroke();
+}
+
+function drawCompleteBadge(ctx) {
+  var s = ctx.canvas.width/100;
+  ctx.beginPath();
+  ctx.arc(s*75, s*75, s*15, 0, Math.TAU, false);
+  ctx.fillStyle = colors.complete;
+  ctx.fill();
+  ctx.strokeStyle = colors.background;
+  ctx.lineWidth = Math.round(s*5);
+  ctx.stroke();
+}
+
+function drawIcon(side, options) {
+  var canvas = document.createElement('canvas');
+  canvas.width = canvas.height = side;
+  document.body.appendChild(canvas);
+  var ctx = canvas.getContext('2d');
+  if (options.anyInProgress) {
+    if (options.anyMissingTotalBytes) {
+      drawUnknownProgressSpinner(ctx);
+    } else {
+      drawProgressSpinner(ctx, (options.totalBytesReceived /
+                                options.totalTotalBytes));
+    }
+  }
+  drawArrow(ctx);
+  if (options.anyDangerous) {
+    drawDangerBadge(ctx);
+  } else if (options.anyPaused) {
+    drawPausedBadge(ctx);
+  } else if (options.anyRecentlyCompleted) {
+    drawCompleteBadge(ctx);
+  }
+  return canvas;
+}
+
+function maybeOpen(id) {
+  var openWhenComplete = [];
+  try {
+    openWhenComplete = JSON.parse(localStorage.openWhenComplete);
```

Comment:
localStorage isn't available in MV3. We'll need to update this before we can test the extension.